src/main/java/org/openstreetmap/atlas/streaming/resource/File.java

```diff
@@ -2,155 +2,485 @@
+                        this.path.getParent());
+            }
+            else
+            {
+                createDirectoriesForPath(this.path.getParent());
+            }
         }
     }
 
     /**
-     * Create a {@link WritableResource} from a {@link java.io.File}.
-     * <p>
-     * Compression is automatically added for the following deflaters:
-     * <ol>
-     * <li>GZIP</li>
-     * </ol>
+     * Create a new {@link File} from a {@link java.io.File}, optionally creating all necessary
+     * parent directories.
      *
-     * @param path
-     *            The path of the file.
+     * @param file
+     *            the {@link java.io.File} to use
+     * @param createParentDirectories
+     *            whether or not to create necessary parent directories
+     * @deprecated please use {@link File#File(Path, boolean)}
      */
-    public File(final String path)
+    @Deprecated
+    public File(final java.io.File file, final boolean createParentDirectories)
     {
-        this(new java.io.File(path), true);
+        this(file.toPath(), createParentDirectories);
     }
 
-    public File(final String path, final boolean createParentDirectories)
+    /**
+     * Create a new {@link File} from a given path string, using the given {@link FileSystem} to
+     * resolve the path string into an actual {@link Path}. Automatically create all necessary
+     * parent directories.
+     *
+     * @param pathString
+     *            the path string to the file
+     * @param fileSystem
+     *            the {@link FileSystem} to use for resolution
+     */
+    public File(final String pathString, final FileSystem fileSystem)
     {
-        this(new java.io.File(path), createParentDirectories);
+        this(pathString, fileSystem, true);
     }
 
+    /**
+     * Create a new {@link File} from a given path string, using the given {@link FileSystem} to
+     * resolve the path string into an actual {@link Path}. Optionally create all necessary parent
+     * directories.
+     *
+     * @param pathString
+     *            the path string to the file
+     * @param fileSystem
+     *            the {@link FileSystem} to use for resolution
+     * @param createParentDirectories
+     *            whether or not to create necessary parent directories
+     */
+    public File(final String pathString, final FileSystem fileSystem,
+            final boolean createParentDirectories)
+    {
+        this(fileSystem.getPath(pathString), createParentDirectories);
+    }
+
+    /**
+     * Create a new {@link File} from a given path string, using the default {@link FileSystem}
+     * (i.e. {@link FileSystems#getDefault()}) to resolve the path string into an actual
+     * {@link Path}. Automatically create all necessary parent directories.
+     *
+     * @param pathString
+     *            the path string to the file
+     */
+    @Deprecated
```

Comment:
Can you add a deprecated comment above, with a description on how to migrate any code using this method to the new model?

Reply:
Done.